scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScannerTokens.scala

```diff
@@ -506,16 +507,72 @@ final class ScannerTokens(val tokens: Tokens)(implicit dialect: Dialect) {
           case xs => xs
         })
       case _: KwFor if dialect.allowSignificantIndentation =>
+        currRef(RegionFor(next) :: sepRegions)
+      case _: KwWhile if dialect.allowSignificantIndentation =>
+        currRef(RegionWhile(next) :: sepRegions)
+      case _: KwIf if dialect.allowSignificantIndentation =>
         currRef(sepRegions match {
-          case (_: RegionParen) :: tail => RegionParen(true) :: tail
-          case _ => sepRegions
+          case (RegionCaseExpr | _: RegionFor) :: _ => sepRegions
+          case (_: RegionDelim) :: (_: RegionFor) :: _ => sepRegions
+          case xs => RegionIf(next) :: xs
         })
+      case _: KwThen if dialect.allowSignificantIndentation =>
+        sepRegions match {
+          case (r: RegionIndent) :: (_: RegionIf) :: xs =>
+            val ref = mkOutdentTo(r, currPos, xs)
+            ref.next = currRef(RegionThen :: xs)
+            ref
+          case (_: RegionIf) :: xs => currRef(RegionThen :: xs)
+          case xs => currRef(RegionThen :: xs)
+        }
+      case _: KwElse if dialect.allowSignificantIndentation =>
+        sepRegions match {
+          case (r: RegionIndent) :: RegionThen :: xs =>
+            val ref = mkOutdentTo(r, currPos, xs)
+            ref.next = currRef(xs)
+            ref
+          case (_: RegionCtrl) :: xs => currRef(xs)
+          case xs => currRef(xs)
+        }
+      case _: KwDo | _: KwYield if dialect.allowSignificantIndentation =>
+        sepRegions match {
+          case (r: RegionIndent) :: (_: RegionCtrl) :: xs =>
+            val ref = mkOutdentTo(r, currPos, xs)
+            ref.next = currRef(xs)
+            ref
+          case (_: RegionCtrl) :: xs => currRef(xs)
+          case xs => currRef(xs)
+        }
       case _ => currRef(sepRegions)
     }
 
     if (currNonTrivial)
       if (isTrailingComma) nextToken(curr, currPos, currPos + 1, sepRegionsOrig)
-      else nonTrivial(sepRegionsOrig)
+      else
+        nonTrivial(sepRegionsOrig match {
```

Comment:
Could you maybe move a body to a separate method?

Reply:
will do